File: sdk/java/src/main/java/io/apibrew/client/ext/Operator.java
package io.apibrew.client.ext;


import io.apibrew.client.ApiException;
import io.apibrew.client.Entity;
import io.apibrew.client.model.Extension;

import java.util.List;
import java.util.Objects;
import java.util.function.*;

import static io.apibrew.client.ext.Condition.*;

public interface Operator<T extends Entity> {

    String operate(Handler<T> handler);

    static <T extends Entity> Operator<T> graceFullDelete(String property, Object value) {

        return handler -> {
            handler = handler.configure(ei -> ei.withResponds(true));
            handler.when(beforeDelete()).operate(graceFullDeleteOperator(property, value));
            handler.when(afterGet()).operate(graceFullDeleteOperator(property, value));
            handler.when(beforeCreate()).operate(graceFullDeleteOperator(property, value));
            handler.when(beforeUpdate()).operate(graceFullDeleteOperator(property, value));

            return handler.when(beforeList()).operate((event, entity) -> {
                Extension.BooleanExpression query = event.getRecordSearchParams().getQuery();

                Extension.BooleanExpression deletedFilterExp = new Extension.BooleanExpression().withNot(
                        new Extension.BooleanExpression().withEqual(
                                new Extension.PairExpression()
                                        .withLeft(new Extension.Expression().withProperty(property))
                                        .withRight(new Extension.Expression().withValue(value))
                        )
                );

                if (query == null) {
                    event.getRecordSearchParams().setQuery(deletedFilterExp);
                } else {
                    event.getRecordSearchParams().setQuery(new Extension.BooleanExpression().withAnd(List.of(query, deletedFilterExp)));
                }

                return entity;
            });
        };
    }

    private static <T extends Entity> BiFunction<Extension.Event, T, T> graceFullDeleteOperator(String property, Object value) {
        return (event, entity) -> {
            if (entity == null) {
                return null;
            }
            if (Objects.equals(entity.getProperty(property), value)) {
                throw new ApiException(Extension.Code.RECORD_VALIDATION_ERROR, "Entity is already deleted");
            } else {
                return entity;
            }
        };
    }

    static <T extends Entity> Operator<T> dataSeparation(String property, Function<T, String> ownerFieldGetter) {
        BiFunction<Extension.Event, T, T> dataSeparator = (event, entity) -> {
            if (entity == null) {
                return null;
            }
            if (Objects.equals(ownerFieldGetter.apply(entity), event.getAnnotations().get("user")))
                return entity;
            else
                throw new ApiException(Extension.Code.RECORD_VALIDATION_ERROR, "Entity is not owned by user");
        };

        return handler -> {
            handler = handler.configure(ei -> ei.withResponds(true));
            handler.when(beforeCreate()).operate(dataSeparator);
            handler.when(beforeUpdate()).operate(dataSeparator);
            handler.when(beforeDelete()).operate(dataSeparator); // fix delete
            handler.when(afterGet()).operate(dataSeparator);

            return handler.when(beforeList()).operate((event, entity) -> {
                Extension.BooleanExpression query = event.getRecordSearchParams().getQuery();

                Extension.BooleanExpression ownerFilterExp = new Extension.BooleanExpression().withEqual(
                        new Extension.PairExpression()
                                .withLeft(new Extension.Expression().withProperty(property))
                                .withRight(new Extension.Expression().withValue(event.getAnnotations().get("user")))
                );

                if (query == null) {
                    event.getRecordSearchParams().setQuery(ownerFilterExp);
                } else {
                    event.getRecordSearchParams().setQuery(new Extension.BooleanExpression().withAnd(List.of(query, ownerFilterExp)));
                }

                return entity;
            });
        };
    }

    static <T extends Entity> Operator<T> execute(BiConsumer<Extension.Event, T> consumer) {
        return handler -> handler.operate((event, entity) -> {
            consumer.accept(event, entity);

            return entity;
        });
    }

    static <T extends Entity> Operator<T> execute(BiFunction<Extension.Event, T, T> consumer) {
        return handler -> handler.operate(consumer);
    }

    static <T extends Entity> Operator<T> reject(String message) {
        return check((event, entity) -> false, message);
    }

    static <T extends Entity> Operator<T> reject() {
        return reject("Operation rejected");
    }

    static <T extends Entity> Operator<T> check(BiPredicate<Extension.Event, T> condition) {
        return check(condition, "Condition not met");
    }

    static <T extends Entity> Operator<T> check(Predicate<T> condition) {
        return check((t, e) -> condition.test(e), "Condition not met");
    }

    static <T extends Entity> Operator<T> check(Predicate<T> condition, String failMessage) {
        return check((t, e) -> condition.test(e), failMessage);
    }

    static <T extends Entity> Operator<T> check(BiPredicate<Extension.Event, T> condition, String failMessage) {
        return handler -> handler.operate((event, entity) -> {
            if (condition.test(event, entity)) {
                return entity;
            } else {
                throw new ApiException(Extension.Code.RECORD_VALIDATION_ERROR, failMessage);
            }
        });
    }
}

File: sdk/nodejs/src/ext/operator.ts
import {Handler} from "./handler";
import {Entity} from "../entity";
import {beforeCreate, beforeDelete, beforeList, beforeUpdate, afterGet, and} from "./condition";
import {Code, Event} from "../model/extension";
import {ApiException} from "../api-exception";
import {BooleanExpressionBuilder} from "../boolean-expression-builder";

export interface Operator<T extends Entity> {
    operate(handler: Handler<T>): string;
}

export function graceFullDelete<T extends Entity>(property: string, value: any): Operator<T> {
    return {
        operate(handler: Handler<T>): string {
            handler = handler.configure(ei => {
                return {
                    ...ei,
                    responds: true,
                }
            });
            handler.when(beforeDelete()).localOperator(graceFullDeleteOperator(property, value));
            handler.when(afterGet()).localOperator(graceFullDeleteOperator(property, value));
            handler.when(beforeCreate()).localOperator(graceFullDeleteOperator(property, value));
            handler.when(beforeUpdate()).localOperator(graceFullDeleteOperator(property, value));

            return handler.when(beforeList()).localOperator((event, entity) => {
                const query = event.recordSearchParams.query;

                const deletedFilterExp = BooleanExpressionBuilder.not(
                    BooleanExpressionBuilder.eq(property, value)
                );

                if (query == null) {
                    event.recordSearchParams.query = deletedFilterExp
                } else {
                    event.recordSearchParams.query = BooleanExpressionBuilder.and(query, deletedFilterExp)
                }

                return entity;
            });

        }
    };
}

export function graceFullDeleteOperator<T extends Entity>(property: string, value: any) {
    return (event: Event, entity: T) => {
        if ((entity as any)[property] === value) {
            throw new ApiException(Code.RECORD_VALIDATION_ERROR, "Entity is already deleted");
        }

        return entity;
    };
}

export function dataSeparation<T extends Entity>(property: string, ownerFieldGetter: (entity: T) => string): Operator<T> {
    const dataSeparator = (event: Event, entity: T) => {
        if ((entity as any)[property] === event.annotations.user) {
            return entity;
        } else {
            throw new ApiException(Code.RECORD_VALIDATION_ERROR, "Entity is not owned by user");
        }
    };

    return {
        operate(handler: Handler<T>): string {
            handler = handler.configure(ei => {
                return {
                    ...ei,
                    responds: true,
                }
            });
            handler.when(beforeCreate()).localOperator(dataSeparator);
            handler.when(beforeUpdate()).localOperator(dataSeparator);
            handler.when(beforeDelete()).localOperator(dataSeparator); // fix delete
            handler.when(afterGet()).localOperator(dataSeparator);

            return handler.when(beforeList()).localOperator((event, entity) => {
                const query = event.recordSearchParams.query;

                const ownerFilterExp = BooleanExpressionBuilder.eq(property, event.annotations.user);

                if (query == null) {
                    event.recordSearchParams.query = ownerFilterExp
                } else {
                    event.recordSearchParams.query = BooleanExpressionBuilder.and(query, ownerFilterExp)
                }

                return entity;
            });

        }
    };
}


export function execute<T extends Entity>(consumer: (event: Event, entity: T) => T | void): Operator<T> {
    return {
        operate(handler: Handler<T>): string {
            return handler.localOperator((event, entity) => {
                consumer(event, entity);

                return entity;
            });
        }
    };
}

export function reject<T extends Entity>(message?: string): Operator<T> {
    return check((event, entity) => false, message);
}

export function check<T extends Entity>(condition: (event: Event, entity: T) => boolean, message?: string): Operator<T> {
    return {
        operate(handler: Handler<T>): string {
            return handler.localOperator((event, entity) => {
                if (condition(event, entity)) {
                    return entity;
                } else {
                    throw new ApiException(Code.RECORD_VALIDATION_ERROR, message || "Condition not met");
                }
            });
        }
    };
}
